Optimize multi_tensor_apply

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119652
* __->__ #119153

### Summary

Due to the dynamic nature of the workload, the kernel arguments aren't guaranteed to fit in the static 4kb kernel argument memory. Previously with the apex implementation, we overcame this limitation by dividing a multi_tensor_apply workload into multiple kernel launches. However, this led to low sustained occupancy, affecting the performance of memory bound ops.

Based on the observation that the kernel argument memory limitation doesn't correlate well with available SM resources, we adopt a different approach:
- When the kernel arguments fit into the static kernel argument memory, we use this memory to transfer the arguments.
- Conversely, when the kernel arguments don't fit into the static kernel argument memory, instead of sacrificing sustained occupancy, we use a page-locked cudaMemcpyAsync to transfer the arguments, then perform the entire workload in a single kernel.

This PR only covers `multi_tensor_apply` for tensors. The change can be easily applied to `multi_tensor_apply` for tensors + scalars and `multi_tensor_apply_for_fused_optimizer`.

### Benchmark (WIP)

The only benchmark I've conducted so far on `_foreach_copy_` on a set of sizes that resembles internal workload. I need to benchmarks on more problem sizes. The speedup should vary among problem sizes. **However, I believe this PR should not be slower than the previous impl on any problem sizes.**

The benchmark can be reproduced with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa).

**Baseline**

A single iteration in trace:
<img width="831" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/5c8d72d0-0628-4989-88a8-c756f6bc1319">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_5a59145f-567b-472f-8eef-c61c388d45b4.json
device ms: 1.111, cpu ms: 7.151
memory bandwidth: 1169.825 GB/s
```

**This PR**

A single iteration in trace:
<img width="967" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/a023e183-8166-48f7-b7c0-c8ba32653d2b">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_da060725-62a8-466e-b570-2ad67ff0e29d.json
device ms: 0.892, cpu ms: 0.810
memory bandwidth: 1456.744 GB/s
```

Commit:
multi-tensor-apply

[ghstack-poisoned]

yifuwang commented:
cc @mlazos @jansel - this may also be applicable to compiled foreach kernels.

Commit:
Update on "Optimize multi_tensor_apply"


### Summary

Due to the dynamic nature of the workload, the kernel arguments aren't guaranteed to fit in the static 4kb kernel argument memory. Previously with the apex implementation, we overcame this limitation by dividing a multi_tensor_apply workload into multiple kernel launches. However, this led to low sustained occupancy, affecting the performance of memory bound ops.

Based on the observation that the kernel argument memory limitation doesn't correlate well with available SM resources, we adopt a different approach:
- When the kernel arguments fit into the static kernel argument memory, we use this memory to transfer the arguments.
- Conversely, when the kernel arguments don't fit into the static kernel argument memory, instead of sacrificing sustained occupancy, we use a page-locked cudaMemcpyAsync to transfer the arguments, then perform the entire workload in a single kernel.

This PR only covers `multi_tensor_apply` for tensors. The change can be easily applied to `multi_tensor_apply` for tensors + scalars and `multi_tensor_apply_for_fused_optimizer`.

### Benchmark (WIP)

The only benchmark I've conducted so far on `_foreach_copy_` on a set of sizes that resembles internal workload. I need to benchmarks on more problem sizes. The speedup should vary among problem sizes. **However, I believe this PR should not be slower than the previous impl on any problem sizes.**

The benchmark can be reproduced with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa).

**Baseline**

A single iteration in trace:
<img width="831" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/5c8d72d0-0628-4989-88a8-c756f6bc1319">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_5a59145f-567b-472f-8eef-c61c388d45b4.json
device ms: 1.111, cpu ms: 7.151
memory bandwidth: 1169.825 GB/s
```

**This PR**

A single iteration in trace:
<img width="967" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/a023e183-8166-48f7-b7c0-c8ba32653d2b">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_da060725-62a8-466e-b570-2ad67ff0e29d.json
device ms: 0.892, cpu ms: 0.810
memory bandwidth: 1456.744 GB/s
```

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -294,6 +341,113 @@ void multi_tensor_apply(
+  const size_t n_tensors = tensor_lists[0].size();
+
+  TORCH_CHECK(
+      n_tensors < static_cast<size_t>(std::numeric_limits<int>::max()),
```
Why "int" is the type checked here? Use a specific bitwidth to ensure it is consistent across os.

yifuwang replied:
Good catch. I updated the type of `block_to_tensor` to `int*` then `uint32_t*` and forgot to update here. I should just go with `size_t` here to eliminate the numeric_limit checks.

albanD commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -294,6 +341,113 @@ void multi_tensor_apply(
+    }
+
+    const auto numel = tensor_lists[0][t].numel();
+    const auto chunks = numel / kChunkSize + (numel % kChunkSize != 0);
```
Same -1 +1 as above

albanD commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -294,6 +341,113 @@ void multi_tensor_apply(
+        block_to_tensor.size() <
+            static_cast<size_t>(std::numeric_limits<uint32_t>::max() - chunks),
+        "multi_tensor_apply: exceeded the maximum supported number of chunks: ",
+        std::numeric_limits<int>::max());
```
raw int

drisspg commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
+  int64_t num_blocks = 0;
+  for (const auto t : c10::irange(n_tensors)) {
+    const auto numel = tensor_lists[0][t].numel();
+    const auto chunks = numel / kChunkSize + (numel % kChunkSize != 0);
```
https://stackoverflow.com/questions/2745074/fast-ceiling-of-an-integer-division-in-c-c

After reading alot of this, I disagree with the suggestion lol

albanD replied:
godbolt doesn't seem to agree though: https://godbolt.org/z/1TMfEhM67

awgu replied:
At this point in the code, have the 0 numel tensors been filtered? If I am not mistaken, it seems like not. In that case, @albanD 's approach has `chunks = 1`, whereas @yifuwang's approach has `chunks = 0` (https://godbolt.org/z/o8a5vfqGh).

drisspg replied:
Andrew points out a good semantic difference but in terms of perf on positive ints
https://quick-bench.com/q/G7tjb5ZDp1Z6MXn0WVYgy4Dz3Vk
I stand mistaken and admit defeat

Anyways we should have a ceil_div function in c10

janeyx99 replied:
Wait @drisspg isn't the second bar slowest in your quick-bench?

awgu replied:
It is worth checking the non-divisible case too: https://quick-bench.com/q/ZveVJxeF6oBfmrux8T-Ajadn6rQ

drisspg replied:
OHHHHH shig  you are right @janeyx99... never mind I claim victory on all my foes!!!

janeyx99 replied:
LOL @awgu  throwing in a wrench --> the nondivisible case is unfortunately higher I would guess.

awgu replied:
I think that one run was an anomaly. I re-ran it a few times, and the time came back down to normal. I cannot seem to reproduce the much slower `div_ceil1`.

Either way, `div_ceil1` handles dividing 0 in the correct way, so maybe we should just keep it as is in the PR.

drisspg replied:
https://quick-bench.com/q/LAg-v-n4HlJ2DfrQsfDdZROV-HE

drisspg replied:
So it turns out that we do have a ceil_div in ATen: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/ceil_div.h

The question now, is this the correct implementation?

janeyx99 replied:
Hhahahaha that's what I think I started with before @r-barnes told me it had risk of overflow and so I corrected it to be the current state it is.

albanD replied:
btw I do think a lot of the benchmark shared here were not super meaningful as all the variables were optimized away.
https://quick-bench.com/q/3ldZe15ZJerQ6oYrmw9AcctBM5w should have something more meaningful but it is very very noisy still.

I would say that it seems to agree with the implementation in ceil_div.h is usually faster.
Also from checking the assembly, it is finding out by itself that it can do a bitshift instead of a full division when the second argument is constant (which it is in our case).

yifuwang replied:
Switched to `at::ceil_div` for now. We can move the battlefield there :)

janeyx99 commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
+  // Copy all previously added vectors onto device and return their device
+  // pointers in the order they are added. We leverage the stream awareness of
+  // CUDACachingAllocator to manage the lifetime of the device arguments - the
+  // device memory is guaranteed to be alive as long as VecPacker is destoryed
```
```suggestion
  // device memory is guaranteed to be alive as long as VecPacker is destroyed
```

janeyx99 commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -294,6 +341,113 @@ void multi_tensor_apply(
+
+    const auto numel = tensor_lists[0][t].numel();
+    const auto chunks = numel / kChunkSize + (numel % kChunkSize != 0);
+    // Using float32 as example, hitting this limit requires rougly 2^32 *
```
```suggestion
    // Using float32 as example, hitting this limit requires roughly 2^32 *
```
another nit

But real q: do we have precedence of setting a "max number of tensors dealt with at time" anywhere else?

yifuwang replied:
I've not seen any intentional cases. Maybe for peace of mind we should just switch to `size_t` here, since we are dynamically allocating kernel arguments and the extra size wouldn't really matter perf-wise.

janeyx99 replied:
Does `size_t` have the same unstandardized bitwidth problem as `int` across different OS's? If not, then sounds good to me.

eqy commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
+        {packed_numel},
+        at::TensorOptions().dtype(at::kByte).pinned_memory(true));
+    for (const auto i : c10::irange(ptrs.size())) {
+      memcpy(packed.data_ptr<uint8_t>() + offsets[i], ptrs[i], sizes[i]);
```
Is it preferred to use `memcpy` or `std::copy` in this case?

eqy commented on aten/src/ATen/native/cuda/ForeachFunctors.cuh:
```diff
@@ -18,7 +18,7 @@ inline void increment_version(TensorList tensors) {
 
 // Initializes args and checks if all args are aligned
-template <int depth, typename T>
+template <int depth, typename T, template <int> class TensorListMetadata>
```
Since we are presumably increasing the amount of specialization, is there any estimate of how much this will increase the binary size per compute-capability?

yifuwang replied:
Good call. I think this would double the size taken by foreach functions, which sounds bad.

It's possible to avoid the specialization, but I'm not sure if it's possible to do so without compromising the fast-path (e.g. if we use the same integer types in both the static and dynamic arg struct, we'd be decreasing the capacity of the static arg struct).

I'll measure the size impact and think more about this.

eqy replied:
Not saying that this should be a blocker for the PR, but it would be useful to know (especially for future changes or updates to this code) what impact increased specialization would have

yifuwang replied:
I measured the binary size. It was 219MB -> 224MB (a 2.28% increase).

I reworked the small buffer optimization mechansim in #119764 so the optimization in this PR can be applied without increasing binary size. I'd appreciate it if you can take a look when you get a chance.

yifuwang commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
+  template <typename T>
+  // Add a vector to be copied to device
+  // NOTE: VecPacker doesn't make copies of the added vectors. They have to be
+  // kept alive by the caller until .pack() is called.
```
Pretty easy I'd say. Since the caller in this context is always our host function that launches cuda kernels.

albanD replied:
But this one is just about the vector<>
In any case, for the use where this is use in one big straight line function. It shouldn't be an issue idd

yifuwang commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
+  void add(const std::vector<T>& vec) {
+    ptrs.push_back(vec.data());
+    const auto vec_bytes = sizeof(T) * vec.size();
+    const auto vec_bytes_aligned = (vec_bytes + 16 - 1) / 16 * 16;
```
Say I add a `std::vector<uint8_t>` of size 7, followed by a `std::vector<int64_t>`. Without alignment, the corresponding dev ptr for the `std::vector<int64_t>` would be `[allocator_alignment] + 7`, which is not aligned with `int64_t` and would lead to unaligned address error.

16 (128-bit) is just something that would cover any T here (I should probably add a `static_assert` here)

janeyx99 replied:
I think you could also use the c10 round_up  (in the same file that @drisspg linked above) here?

albanD replied:
Yeah that would be great to mention that c10::complex<double> is the reason why we need 16 here and a static assert that no local type is bigger than that would be great.

janeyx99 commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
+}
+
+// Helper for transfering multiple std::vector<T> onto device with a single
+// page-locked cudaMemcpyAsync.
```
For my learning: where is the page locking happening? And the cudaMemcpyAsync? I see a `memcpy` and a `.to` call below--do one of them implicitly call cudaMemcpyAsync?

yifuwang replied:
> Update: oooooh is the page-locked referring to the .pinned_memory call below?

Yes this is correct. We get the pinned (a.k.a page-locked) memory from `CachingHostAllocator` so it's practically free. When the src is in pinned memory, `.to(/*non_blocking=*/true)` results in a `cudaMemcpyAsync`.

janeyx99 commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
+    return dev_ptrs;
+  }
+
+  std::vector<const void*> ptrs;
+  std::vector<size_t> sizes;
+  std::vector<size_t> offsets;
+  int64_t packed_numel = 0;
+  at::Tensor packed;
```
Style q: is it typical to put the struct attributes beneath the function defs? I think I'm more used to these being above cuz I find it easier to read top-down but I don't actually know the style recommendation.

yifuwang replied:
I'm not familiar with any. But now that you mention it, it does look better to put them on top :)

janeyx99 commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -294,6 +341,113 @@ void multi_tensor_apply(
+
+  for (size_t t = 0; t < n_tensors; t++) {
+    // short-circuit to avoid adding empty tensors to tensorListMeta
+    if (tensor_lists[0][t].numel() == 0) {
+      continue;
+    }
+    numel_for_tensor.push_back(tensor_lists[0][t].numel());
+    for (int d = 0; d < depth; d++) {
+      addresses[d].push_back(tensor_lists[d][t].const_data_ptr());
+    }
+
+    const auto numel = tensor_lists[0][t].numel();
```
```suggestion
    const auto numel = tensor_lists[0][t].numel();
    if (numel == 0) {
      continue;
    }
    numel_for_tensor.push_back(numel);
    for (int d = 0; d < depth; d++) {
      addresses[d].push_back(tensor_lists[d][t].const_data_ptr());
    }

```

janeyx99 reviewed (commented):
This change looks pretty good to me. 

I'm curious if there's _ever_ a feasible case where we'd need more than one kernel launch; i.e., is there a limit where the pagelocked memcpy will be slow?

I'm also curious to see how this will abstract across the three mtas vs what's done today with lots of copy/paste 😛 .

Commit:
Update on "Optimize multi_tensor_apply"


### Summary

Due to the dynamic nature of the workload, the kernel arguments aren't guaranteed to fit in the static 4kb kernel argument memory. Previously with the apex implementation, we overcame this limitation by dividing a multi_tensor_apply workload into multiple kernel launches. However, this led to low sustained occupancy, affecting the performance of memory bound ops.

Based on the observation that the kernel argument memory limitation doesn't correlate well with available SM resources, we adopt a different approach:
- When the kernel arguments fit into the static kernel argument memory, we use this memory to transfer the arguments.
- Conversely, when the kernel arguments don't fit into the static kernel argument memory, instead of sacrificing sustained occupancy, we use a page-locked cudaMemcpyAsync to transfer the arguments, then perform the entire workload in a single kernel.

This PR only covers `multi_tensor_apply` for tensors. The change can be easily applied to `multi_tensor_apply` for tensors + scalars and `multi_tensor_apply_for_fused_optimizer`.

### Benchmark (WIP)

The only benchmark I've conducted so far on `_foreach_copy_` on a set of sizes that resembles internal workload. I need to benchmarks on more problem sizes. The speedup should vary among problem sizes. **However, I believe this PR should not be slower than the previous impl on any problem sizes.**

The benchmark can be reproduced with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa).

**Baseline**

A single iteration in trace:
<img width="831" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/5c8d72d0-0628-4989-88a8-c756f6bc1319">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_5a59145f-567b-472f-8eef-c61c388d45b4.json
device ms: 1.111, cpu ms: 7.151
memory bandwidth: 1169.825 GB/s
```

**This PR**

A single iteration in trace:
<img width="967" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/a023e183-8166-48f7-b7c0-c8ba32653d2b">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_da060725-62a8-466e-b570-2ad67ff0e29d.json
device ms: 0.892, cpu ms: 0.810
memory bandwidth: 1456.744 GB/s
```

[ghstack-poisoned]

janeyx99 reviewed (approved):
Looks good to me! Thanks so much for this change c:

Do you think you could enable this for the other two multitensorapply functions?

Would be good to measure the bin size increase but not a blocker here.

I’m not an expert of c++ perf so feel free to get others’ thoughts on std::copy vs memcpy.

Commit:
Update on "Optimize multi_tensor_apply"


### Summary

Due to the dynamic nature of the workload, the kernel arguments aren't guaranteed to fit in the static 4kb kernel argument memory. Previously with the apex implementation, we overcame this limitation by dividing a multi_tensor_apply workload into multiple kernel launches. However, this led to low sustained occupancy, affecting the performance of memory bound ops.

Based on the observation that the kernel argument memory limitation doesn't correlate well with available SM resources, we adopt a different approach:
- When the kernel arguments fit into the static kernel argument memory, we use this memory to transfer the arguments.
- Conversely, when the kernel arguments don't fit into the static kernel argument memory, instead of sacrificing sustained occupancy, we use a page-locked cudaMemcpyAsync to transfer the arguments, then perform the entire workload in a single kernel.

This PR only covers `multi_tensor_apply` for tensors. The change can be easily applied to `multi_tensor_apply` for tensors + scalars and `multi_tensor_apply_for_fused_optimizer`.

### Benchmark (WIP)

The only benchmark I've conducted so far on `_foreach_copy_` on a set of sizes that resembles internal workload. I need to benchmarks on more problem sizes. The speedup should vary among problem sizes. **However, I believe this PR should not be slower than the previous impl on any problem sizes.**

The benchmark can be reproduced with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa).

**Baseline**

A single iteration in trace:
<img width="831" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/5c8d72d0-0628-4989-88a8-c756f6bc1319">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_5a59145f-567b-472f-8eef-c61c388d45b4.json
device ms: 1.111, cpu ms: 7.151
memory bandwidth: 1169.825 GB/s
```

**This PR**

A single iteration in trace:
<img width="967" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/a023e183-8166-48f7-b7c0-c8ba32653d2b">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_da060725-62a8-466e-b570-2ad67ff0e29d.json
device ms: 0.892, cpu ms: 0.810
memory bandwidth: 1456.744 GB/s
```

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

Commit:
Update on "Optimize multi_tensor_apply"


### Summary

Due to the dynamic nature of the workload, the kernel arguments aren't guaranteed to fit in the static 4kb kernel argument memory. Previously with the apex implementation, we overcame this limitation by dividing a multi_tensor_apply workload into multiple kernel launches. However, this led to low sustained occupancy, affecting the performance of memory bound ops.

Based on the observation that the kernel argument memory limitation doesn't correlate well with available SM resources, we adopt a different approach:
- When the kernel arguments fit into the static kernel argument memory, we use this memory to transfer the arguments.
- Conversely, when the kernel arguments don't fit into the static kernel argument memory, instead of sacrificing sustained occupancy, we use a page-locked cudaMemcpyAsync to transfer the arguments, then perform the entire workload in a single kernel.

This PR only covers `multi_tensor_apply` for tensors. The change can be easily applied to `multi_tensor_apply` for tensors + scalars and `multi_tensor_apply_for_fused_optimizer`.

### Benchmark (WIP)

The only benchmark I've conducted so far on `_foreach_copy_` on a set of sizes that resembles internal workload. I need to benchmarks on more problem sizes. The speedup should vary among problem sizes. **However, I believe this PR should not be slower than the previous impl on any problem sizes.**

The benchmark can be reproduced with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa).

**Baseline**

A single iteration in trace:
<img width="831" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/5c8d72d0-0628-4989-88a8-c756f6bc1319">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_5a59145f-567b-472f-8eef-c61c388d45b4.json
device ms: 1.111, cpu ms: 7.151
memory bandwidth: 1169.825 GB/s
```

**This PR**

A single iteration in trace:
<img width="967" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/a023e183-8166-48f7-b7c0-c8ba32653d2b">

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_da060725-62a8-466e-b570-2ad67ff0e29d.json
device ms: 0.892, cpu ms: 0.810
memory bandwidth: 1456.744 GB/s
```

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

albanD commented:
@yifuwang this is actually breaking cudagraph (since the copy from host to device is now allowed when capturing).
I think we need to add a check that checks if we are capturing a graph and if so, go back to the old behavior where it launched multiple kernels without every doing a host to device copy.

yifuwang commented:
Thanks for the heads up @albanD. I'll revert this first.

@pytorchbot revert this please

yifuwang commented:
@pytorchbot revert -m "breaking cuda graph"

yifuwang commented:
@pytorchbot revert -m="This PR is breaking cuda graph for multi_tensor_apply" -c=nosignal

yifuwang commented:
Since I had a PR (#119652) that eliminates the binary size increase introduced by this PR, and this PR was reverted due to a CUDA graph issue, I created a new PR that combines the two + the fix for the CUDA graph issue: https://github.com/pytorch/pytorch/pull/119764
